File: packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.sharedpreferences;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.mockito.Mockito.anyInt;
import static org.mockito.Mockito.anyString;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.Mockito;

public class SharedPreferencesTest {

  public static class LocalSharedPreferencesEditor implements SharedPreferences.Editor {
    private final Map<String, Object> sharedPrefData;

    LocalSharedPreferencesEditor(@NonNull Map<String, Object> data) {
      sharedPrefData = data;
    }

    @Override
    public @NonNull SharedPreferences.Editor putString(@NonNull String key, @NonNull String value) {
      sharedPrefData.put(key, value);
      return this;
    }

    @Override
    public @NonNull SharedPreferences.Editor putStringSet(
        @NonNull String key, @NonNull Set<String> values) {
      sharedPrefData.put(key, values);
      return this;
    }

    @Override
    public @NonNull SharedPreferences.Editor putBoolean(
        @NonNull String key, @NonNull boolean value) {
      sharedPrefData.put(key, value);
      return this;
    }

    @Override
    public @NonNull SharedPreferences.Editor putInt(@NonNull String key, @NonNull int value) {
      sharedPrefData.put(key, value);
      return this;
    }

    @Override
    public @NonNull SharedPreferences.Editor putLong(@NonNull String key, @NonNull long value) {
      sharedPrefData.put(key, value);
      return this;
    }

    @Override
    public @NonNull SharedPreferences.Editor putFloat(@NonNull String key, @NonNull float value) {
      sharedPrefData.put(key, value);
      return this;
    }

    @Override
    public @NonNull SharedPreferences.Editor remove(@NonNull String key) {
      sharedPrefData.remove(key);
      return this;
    }

    @Override
    public @NonNull boolean commit() {
      return true;
    }

    @Override
    public void apply() {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public @NonNull SharedPreferences.Editor clear() {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }
  }

  private static class LocalSharedPreferences implements SharedPreferences {

    static Map<String, Object> sharedPrefData = new HashMap<>();

    @Override
    public @NonNull Map<String, ?> getAll() {
      return sharedPrefData;
    }

    @Override
    public @NonNull SharedPreferences.Editor edit() {
      return new LocalSharedPreferencesEditor(sharedPrefData);
    }

    // All methods below are not implemented.
    @Override
    public @NonNull boolean contains(@NonNull String key) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public @NonNull boolean getBoolean(@NonNull String key, @NonNull boolean defValue) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public @NonNull float getFloat(@NonNull String key, @NonNull float defValue) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public @NonNull int getInt(@NonNull String key, @NonNull int defValue) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public @NonNull long getLong(@NonNull String key, @NonNull long defValue) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public @NonNull String getString(@NonNull String key, @NonNull String defValue) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public @NonNull Set<String> getStringSet(@NonNull String key, @NonNull Set<String> defValues) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public void registerOnSharedPreferenceChangeListener(
        @NonNull SharedPreferences.OnSharedPreferenceChangeListener listener) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }

    @Override
    public void unregisterOnSharedPreferenceChangeListener(
        @NonNull SharedPreferences.OnSharedPreferenceChangeListener listener) {
      throw new UnsupportedOperationException("This method is not implemented for testing");
    }
  }

  static class ListEncoder implements SharedPreferencesListEncoder {
    @Override
    public @NonNull String encode(@NonNull List<String> list) {
      return list.toString();
    }

    @Override
    public @NonNull List<String> decode(@NonNull String listString) {
      return Arrays.asList(listString.substring(1, listString.length() - 1).split(", "));
    }
  }

  SharedPreferencesPlugin plugin;

  @Mock BinaryMessenger mockMessenger;
  @Mock FlutterPlugin.FlutterPluginBinding flutterPluginBinding;

  @Before
  public void before() {
    Context context = Mockito.mock(Context.class);
    SharedPreferences sharedPrefs = new LocalSharedPreferences();

    flutterPluginBinding = Mockito.mock(FlutterPlugin.FlutterPluginBinding.class);

    Mockito.when(flutterPluginBinding.getBinaryMessenger()).thenReturn(mockMessenger);
    Mockito.when(flutterPluginBinding.getApplicationContext()).thenReturn(context);
    Mockito.when(context.getSharedPreferences(anyString(), anyInt())).thenReturn(sharedPrefs);

    plugin = new SharedPreferencesPlugin(new ListEncoder());
    plugin.onAttachedToEngine(flutterPluginBinding);
  }

  private static final Map<String, Object> data = new HashMap<>();

  static {
    data.put("Language", "Java");
    data.put("Counter", 0L);
    data.put("Pie", 3.14);
    data.put("Names", Arrays.asList("Flutter", "Dart"));
    data.put("NewToFlutter", false);
    data.put("flutter.Language", "Java");
    data.put("flutter.Counter", 0L);
    data.put("flutter.Pie", 3.14);
    data.put("flutter.Names", Arrays.asList("Flutter", "Dart"));
    data.put("flutter.NewToFlutter", false);
    data.put("prefix.Language", "Java");
    data.put("prefix.Counter", 0L);
    data.put("prefix.Pie", 3.14);
    data.put("prefix.Names", Arrays.asList("Flutter", "Dart"));
    data.put("prefix.NewToFlutter", false);
  }

  @Test
  public void getAllWithPrefix() {
    plugin.clearWithPrefix("");

    assertEquals(plugin.getAllWithPrefix("").size(), 0);

    addData();

    Map<String, Object> flutterData = plugin.getAllWithPrefix("flutter.");

    assertEquals(flutterData.size(), 5);
    assertEquals(flutterData.get("flutter.Language"), "Java");
    assertEquals(flutterData.get("flutter.Counter"), 0L);
    assertEquals(flutterData.get("flutter.Pie"), 3.14);
    assertEquals(flutterData.get("flutter.Names"), Arrays.asList("Flutter", "Dart"));
    assertEquals(flutterData.get("flutter.NewToFlutter"), false);

    Map<String, Object> allData = plugin.getAllWithPrefix("");

    assertEquals(allData, data);
  }

  @Test
  public void clearWithPrefix() {
    plugin.clearWithPrefix("");

    addData();

    assertEquals(plugin.getAllWithPrefix("").size(), 15);

    plugin.clearWithPrefix("flutter.");

    assertEquals(plugin.getAllWithPrefix("").size(), 10);
  }

  @Test
  public void testRemove() {
    plugin.clearWithPrefix("");

    plugin.setBool("isJava", true);
    assert (plugin.getAllWithPrefix("").containsKey("isJava"));
    plugin.remove("isJava");
    assertFalse(plugin.getAllWithPrefix("").containsKey("isJava"));
  }

  private void addData() {
    plugin.setString("Language", "Java");
    plugin.setInt("Counter", 0);
    plugin.setDouble("Pie", 3.14);
    plugin.setStringList("Names", Arrays.asList("Flutter", "Dart"));
    plugin.setBool("NewToFlutter", false);
    plugin.setString("flutter.Language", "Java");
    plugin.setInt("flutter.Counter", 0);
    plugin.setDouble("flutter.Pie", 3.14);
    plugin.setStringList("flutter.Names", Arrays.asList("Flutter", "Dart"));
    plugin.setBool("flutter.NewToFlutter", false);
    plugin.setString("prefix.Language", "Java");
    plugin.setInt("prefix.Counter", 0);
    plugin.setDouble("prefix.Pie", 3.14);
    plugin.setStringList("prefix.Names", Arrays.asList("Flutter", "Dart"));
    plugin.setBool("prefix.NewToFlutter", false);
  }
}

File: packages/shared_preferences/shared_preferences_android/android/src/test/kotlin/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.kt
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.sharedpreferences

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import androidx.preference.PreferenceManager
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.mockk.every
import io.mockk.mockk
import java.io.ByteArrayOutputStream
import java.io.ObjectOutputStream
import org.junit.Assert
import org.junit.Assert.assertThrows
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
internal class SharedPreferencesTest {
  private val stringKey = "testString"

  private val boolKey = "testBool"

  private val intKey = "testInt"

  private val doubleKey = "testDouble"

  private val listKey = "testList"

  private val testString = "hello world"

  private val testBool = true

  private val testInt = 42L

  private val testDouble = 3.14159

  private val testList = listOf("foo", "bar")

  private val dataStoreOptions = SharedPreferencesPigeonOptions(useDataStore = true)
  private val sharedPreferencesOptions = SharedPreferencesPigeonOptions(useDataStore = false)
  private val testContext: Context = ApplicationProvider.getApplicationContext()

  private fun pluginSetup(options: SharedPreferencesPigeonOptions): SharedPreferencesAsyncApi {
    val plugin = SharedPreferencesPlugin()
    val binaryMessenger = mockk<BinaryMessenger>()
    val flutterPluginBinding = mockk<FlutterPlugin.FlutterPluginBinding>()
    every { flutterPluginBinding.binaryMessenger } returns binaryMessenger
    every { flutterPluginBinding.applicationContext } returns testContext
    plugin.onAttachedToEngine(flutterPluginBinding)
    val backend =
        SharedPreferencesBackend(
            flutterPluginBinding.binaryMessenger, flutterPluginBinding.applicationContext)
    return if (options.useDataStore) {
      plugin
    } else {
      backend
    }
  }

  @Test
  fun testSetAndGetBoolWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setBool(boolKey, testBool, dataStoreOptions)
    Assert.assertEquals(plugin.getBool(boolKey, dataStoreOptions), testBool)
  }

  @Test
  fun testSetAndGetStringWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setString(stringKey, testString, dataStoreOptions)
    Assert.assertEquals(plugin.getString(stringKey, dataStoreOptions), testString)
  }

  @Test
  fun testSetAndGetIntWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setInt(intKey, testInt, dataStoreOptions)
    Assert.assertEquals(plugin.getInt(intKey, dataStoreOptions), testInt)
  }

  @Test
  fun testSetAndGetDoubleWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setDouble(doubleKey, testDouble, dataStoreOptions)
    Assert.assertEquals(plugin.getDouble(doubleKey, dataStoreOptions), testDouble)
  }

  @Test
  fun testSetAndGetStringListWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setStringList(listKey, testList, dataStoreOptions)
    Assert.assertEquals(plugin.getStringList(listKey, dataStoreOptions), testList)
  }

  @Test
  fun testGetKeysWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setBool(boolKey, testBool, dataStoreOptions)
    plugin.setString(stringKey, testString, dataStoreOptions)
    plugin.setInt(intKey, testInt, dataStoreOptions)
    plugin.setDouble(doubleKey, testDouble, dataStoreOptions)
    plugin.setStringList(listKey, testList, dataStoreOptions)
    val keyList = plugin.getKeys(listOf(boolKey, stringKey), dataStoreOptions)
    Assert.assertEquals(keyList.size, 2)
    Assert.assertTrue(keyList.contains(stringKey))
    Assert.assertTrue(keyList.contains(boolKey))
  }

  @Test
  fun testClearWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setBool(boolKey, testBool, dataStoreOptions)
    plugin.setString(stringKey, testString, dataStoreOptions)
    plugin.setInt(intKey, testInt, dataStoreOptions)
    plugin.setDouble(doubleKey, testDouble, dataStoreOptions)
    plugin.setStringList(listKey, testList, dataStoreOptions)

    plugin.clear(null, dataStoreOptions)

    Assert.assertNull(plugin.getBool(boolKey, dataStoreOptions))
    Assert.assertNull(plugin.getBool(stringKey, dataStoreOptions))
    Assert.assertNull(plugin.getBool(intKey, dataStoreOptions))
    Assert.assertNull(plugin.getBool(doubleKey, dataStoreOptions))
    Assert.assertNull(plugin.getBool(listKey, dataStoreOptions))
  }

  @Test
  fun testGetAllWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setBool(boolKey, testBool, dataStoreOptions)
    plugin.setString(stringKey, testString, dataStoreOptions)
    plugin.setInt(intKey, testInt, dataStoreOptions)
    plugin.setDouble(doubleKey, testDouble, dataStoreOptions)
    plugin.setStringList(listKey, testList, dataStoreOptions)

    val all = plugin.getAll(null, dataStoreOptions)

    Assert.assertEquals(all[boolKey], testBool)
    Assert.assertEquals(all[stringKey], testString)
    Assert.assertEquals(all[intKey], testInt)
    Assert.assertEquals(all[doubleKey], testDouble)
    Assert.assertEquals(all[listKey], testList)
  }

  @Test
  fun testClearWithAllowListWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setBool(boolKey, testBool, dataStoreOptions)
    plugin.setString(stringKey, testString, dataStoreOptions)
    plugin.setInt(intKey, testInt, dataStoreOptions)
    plugin.setDouble(doubleKey, testDouble, dataStoreOptions)
    plugin.setStringList(listKey, testList, dataStoreOptions)

    plugin.clear(listOf(boolKey, stringKey), dataStoreOptions)

    Assert.assertNull(plugin.getBool(boolKey, dataStoreOptions))
    Assert.assertNull(plugin.getString(stringKey, dataStoreOptions))
    Assert.assertNotNull(plugin.getInt(intKey, dataStoreOptions))
    Assert.assertNotNull(plugin.getDouble(doubleKey, dataStoreOptions))
    Assert.assertNotNull(plugin.getStringList(listKey, dataStoreOptions))
  }

  @Test
  fun testGetAllWithAllowListWithDataStore() {
    val plugin = pluginSetup(dataStoreOptions)
    plugin.setBool(boolKey, testBool, dataStoreOptions)
    plugin.setString(stringKey, testString, dataStoreOptions)
    plugin.setInt(intKey, testInt, dataStoreOptions)
    plugin.setDouble(doubleKey, testDouble, dataStoreOptions)
    plugin.setStringList(listKey, testList, dataStoreOptions)

    val all = plugin.getAll(listOf(boolKey, stringKey), dataStoreOptions)

    Assert.assertEquals(all[boolKey], testBool)
    Assert.assertEquals(all[stringKey], testString)
    Assert.assertNull(all[intKey])
    Assert.assertNull(all[doubleKey])
    Assert.assertNull(all[listKey])
  }

  @Test
  fun testSetAndGetBoolWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setBool(boolKey, testBool, sharedPreferencesOptions)
    Assert.assertEquals(plugin.getBool(boolKey, sharedPreferencesOptions), testBool)
  }

  @Test
  fun testSetAndGetStringWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setString(stringKey, testString, sharedPreferencesOptions)
    Assert.assertEquals(plugin.getString(stringKey, sharedPreferencesOptions), testString)
  }

  @Test
  fun testSetAndGetIntWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setInt(intKey, testInt, sharedPreferencesOptions)
    Assert.assertEquals(plugin.getInt(intKey, sharedPreferencesOptions), testInt)
  }

  @Test
  fun testSetAndGetDoubleWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setDouble(doubleKey, testDouble, sharedPreferencesOptions)
    Assert.assertEquals(plugin.getDouble(doubleKey, sharedPreferencesOptions), testDouble)
  }

  @Test
  fun testSetAndGetStringListWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setStringList(listKey, testList, sharedPreferencesOptions)
    Assert.assertEquals(plugin.getStringList(listKey, sharedPreferencesOptions), testList)
  }

  @Test
  fun testGetKeysWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setBool(boolKey, testBool, sharedPreferencesOptions)
    plugin.setString(stringKey, testString, sharedPreferencesOptions)
    plugin.setInt(intKey, testInt, sharedPreferencesOptions)
    plugin.setDouble(doubleKey, testDouble, sharedPreferencesOptions)
    plugin.setStringList(listKey, testList, sharedPreferencesOptions)
    val keyList = plugin.getKeys(listOf(boolKey, stringKey), sharedPreferencesOptions)
    Assert.assertEquals(keyList.size, 2)
    Assert.assertTrue(keyList.contains(stringKey))
    Assert.assertTrue(keyList.contains(boolKey))
  }

  @Test
  fun testClearWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setBool(boolKey, testBool, sharedPreferencesOptions)
    plugin.setString(stringKey, testString, sharedPreferencesOptions)
    plugin.setInt(intKey, testInt, sharedPreferencesOptions)
    plugin.setDouble(doubleKey, testDouble, sharedPreferencesOptions)
    plugin.setStringList(listKey, testList, sharedPreferencesOptions)

    plugin.clear(null, sharedPreferencesOptions)

    Assert.assertNull(plugin.getBool(boolKey, sharedPreferencesOptions))
    Assert.assertNull(plugin.getBool(stringKey, sharedPreferencesOptions))
    Assert.assertNull(plugin.getBool(intKey, sharedPreferencesOptions))
    Assert.assertNull(plugin.getBool(doubleKey, sharedPreferencesOptions))
    Assert.assertNull(plugin.getBool(listKey, sharedPreferencesOptions))
  }

  @Test
  fun testGetAllWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setBool(boolKey, testBool, sharedPreferencesOptions)
    plugin.setString(stringKey, testString, sharedPreferencesOptions)
    plugin.setInt(intKey, testInt, sharedPreferencesOptions)
    plugin.setDouble(doubleKey, testDouble, sharedPreferencesOptions)
    plugin.setStringList(listKey, testList, sharedPreferencesOptions)

    val all = plugin.getAll(null, sharedPreferencesOptions)

    Assert.assertEquals(all[boolKey], testBool)
    Assert.assertEquals(all[stringKey], testString)
    Assert.assertEquals(all[intKey], testInt)
    Assert.assertEquals(all[doubleKey], testDouble)
    Assert.assertEquals(all[listKey], testList)
  }

  @Test
  fun testClearWithAllowListWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setBool(boolKey, testBool, sharedPreferencesOptions)
    plugin.setString(stringKey, testString, sharedPreferencesOptions)
    plugin.setInt(intKey, testInt, sharedPreferencesOptions)
    plugin.setDouble(doubleKey, testDouble, sharedPreferencesOptions)
    plugin.setStringList(listKey, testList, sharedPreferencesOptions)

    plugin.clear(listOf(boolKey, stringKey), sharedPreferencesOptions)

    Assert.assertNull(plugin.getBool(boolKey, sharedPreferencesOptions))
    Assert.assertNull(plugin.getString(stringKey, sharedPreferencesOptions))
    Assert.assertNotNull(plugin.getInt(intKey, sharedPreferencesOptions))
    Assert.assertNotNull(plugin.getDouble(doubleKey, sharedPreferencesOptions))
    Assert.assertNotNull(plugin.getStringList(listKey, sharedPreferencesOptions))
  }

  @Test
  fun testGetAllWithAllowListWithSharedPreferences() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    plugin.setBool(boolKey, testBool, sharedPreferencesOptions)
    plugin.setString(stringKey, testString, sharedPreferencesOptions)
    plugin.setInt(intKey, testInt, sharedPreferencesOptions)
    plugin.setDouble(doubleKey, testDouble, sharedPreferencesOptions)
    plugin.setStringList(listKey, testList, sharedPreferencesOptions)

    val all = plugin.getAll(listOf(boolKey, stringKey), sharedPreferencesOptions)

    Assert.assertEquals(all[boolKey], testBool)
    Assert.assertEquals(all[stringKey], testString)
    Assert.assertNull(all[intKey])
    Assert.assertNull(all[doubleKey])
    Assert.assertNull(all[listKey])
  }

  @Test
  fun testSharedPreferencesWithMultipleFiles() {
    val plugin = pluginSetup(sharedPreferencesOptions)
    val optionsWithNewFile =
        SharedPreferencesPigeonOptions(useDataStore = false, fileName = "test_file")
    plugin.setInt(intKey, 1, sharedPreferencesOptions)
    plugin.setInt(intKey, 2, optionsWithNewFile)
    Assert.assertEquals(plugin.getInt(intKey, sharedPreferencesOptions), 1L)
    Assert.assertEquals(plugin.getInt(intKey, optionsWithNewFile), 2L)
  }

  @Test
  fun testSharedPreferencesDefaultFile() {
    val defaultPreferences: SharedPreferences =
        PreferenceManager.getDefaultSharedPreferences(testContext)
    defaultPreferences.edit().putString(stringKey, testString).commit()
    val plugin = pluginSetup(sharedPreferencesOptions)
    Assert.assertEquals(plugin.getString(stringKey, sharedPreferencesOptions), testString)
  }

  @Test
  fun testUnexpectedClassDecodeThrows() {
    // Only String should be allowed in an encoded list.
    val badList = listOf(1, 2, 3)
    // Replicate the behavior of ListEncoder.encode, but with a non-List<String> list.
    val byteStream = ByteArrayOutputStream()
    val stream = ObjectOutputStream(byteStream)
    stream.writeObject(badList)
    stream.flush()
    val badPref = LIST_PREFIX + Base64.encodeToString(byteStream.toByteArray(), 0)

    val plugin = pluginSetup(dataStoreOptions)
    val badListKey = "badList"
    // Inject the bad pref as a string, as that is how string lists are stored internally.
    plugin.setString(badListKey, badPref, dataStoreOptions)
    assertThrows(ClassNotFoundException::class.java) {
      plugin.getStringList(badListKey, dataStoreOptions)
    }
  }
}
